actionpack/test/controller/parameters/mutators_test.rb

```diff
@@ -217,4 +217,14 @@ class ParametersMutatorsTest < ActiveSupport::TestCase
     params = ActionController::Parameters.new(name: "Alex", age: "40", location: "Beijing")
     assert_raises(ActionController::UnfilteredParameters) { params.to_h { |key, value| [key, value] } }
   end
+
+  test "[]= converts hashes to parameters" do
+    params = ActionController::Parameters.new(name: "Bruce")
+    guitar = { "make" => "Fender", "model" => "Telecaster" }
+    params[:guitar] = guitar
+    assert_equal(
+      ActionController::Parameters.new(guitar),
+      params.instance_variable_get(:@parameters)[:guitar], # bypass #Parameters#[], which also converts hashes to parameters
```

Comment:
Since `[]` converts the value before returning it, this was the only way I could think of to assert that `[]=` does the conversion.

Reply:
As there's a protected `attr_reader` for `@parameters`, I guess you could do

```suggestion
      params.send(:parameters)[:guitar], # bypass #Parameters#[], which also converts hashes to parameters
```

Which is definitely _shorter_. But I appreciate using `send` is a slippery slope. 😄

Reply:
Instead of making the test very specific to `[]=`, how about doing something more similar to the test in your reproduction script?

Reply:
If we only want that, then the test exercising `as_json` should do it, and I can delete this one.